src/Microsoft.OData.Core/ODataInnerError.cs

```diff
@@ -30,17 +32,44 @@ public ODataInnerError()
         /// <summary>Initializes a new instance of the <see cref="T:Microsoft.OData.ODataInnerError" /> class with exception object.</summary>
         /// <param name="exception">The <see cref="T:System.Exception" /> used to create the inner error.</param>
         public ODataInnerError(Exception exception)
+            : this(exception, "internalexception")
         {
-            ExceptionUtils.CheckArgumentNotNull(exception, "exception");
 
-            this.Message = exception.Message ?? string.Empty;
-            this.TypeName = exception.GetType().FullName;
-            this.StackTrace = exception.StackTrace;
+        }
+
+        public ODataInnerError(Exception exception, string nestedObjectName)
+        {
+            ExceptionUtils.CheckArgumentNotNull(exception, "exception");
 
             if (exception.InnerException != null)
             {
-                this.InnerError = new ODataInnerError(exception.InnerException);
+                this.InnerError = new ODataInnerError(exception.InnerException, nestedObjectName);
             }
+
+            Properties = new Dictionary<string, string>();
+            Properties.Add("message", exception.Message ?? string.Empty);
+            Properties.Add("type", exception.GetType().FullName ?? string.Empty);
+            Properties.Add("stacktrace", exception.Message ?? string.Empty);
+        }
+
+        public ODataInnerError(IDictionary<string, string> properties, string nestedObjectName, ODataInnerError nestedObject)

```

Comment:
>string [](start = 51, length = 6)

do dynamic properties have to be string-valued? why not dictionary<string,object>?

Reply:
It could be Dictionary<string, object>. However, the object here should be instance of built in "primitive", enum, resource, and collection. Otherwise, maybe use Dictionary<string, ODataValue>.

---
In reply to: [314483310](https://github.com/OData/odata.net/pull/1523#discussion_r314483310) [](ancestors = 314483310)